refactor: examples and tests

rajatagarwal-ibm opened the pull request:
### Description
* Updated GitHub CI file
* Terradocs for FSCloud profile readme
* Added "autoscale" and "replica" examples to the "complete" example. 
* Removed "autoscale" and "replica" examples.
* Updated pr_test and other_test as per the https://github.ibm.com/GoldenEye/issues/issues/4668#issuecomment-56795084

### Types of changes in this PR

#### No release required

- [x] Examples or tests (addition or updates of examples or tests)
- [ ] Documentation update
- [ ] CI-related update (pipeline, etc.)
- [ ] Other changes that don't affect Terraform code

#### Release required

- [ ] Bug fix (patch release (`x.x.X`): Change that fixes an issue and is compatible with earlier versions)
- [ ] New feature (minor release (`x.X.x`): Change that adds functionality and is compatible with earlier versions)
- [ ] Breaking change (major release (`X.x.x`): Change that is likely incompatible with previous versions)

##### Release notes content

Replace this text with information that users need to know about the bug fixes, features, and breaking changes. This information helps the merger write the commit message that is published in the release notes for the module.

---

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### Merge actions for mergers

- Merge by using "Squash and merge".
- Use a relevant [conventional commit](https://www.conventionalcommits.org/) message that is based on the PR contents and any release notes provided by the PR author.

    The commit message determines whether a new version of the module is needed, and if so, which semver increment to use (major, minor, or patch).

ocofaigh commented on tests/other_test.go:
```diff
@@ -22,3 +24,44 @@ func TestRunRestoredDBExample(t *testing.T) {
+
+func TestRunBasicExample(t *testing.T) {
+	t.Parallel()
+	versions := []string{"12", "13"} // Always test for version which are already not included in the complete and FSCloud tests.
```
Since we actually already have two other examples that we could test these versions with (PITR and restore) then lets lock those into these versions, and pass no version into the basic example so it will always use the current ICD default version

rajatagarwal-ibm replied:
The restore test would create 4 Postgres instances (primary and backup * 2) rather than just 2 if we test against 2 different versions. I think we can keep as is because the purpose of having different versions tested is to make sure about things are working fine - that can be tested in the basic example as well.

ocofaigh reviewed (changes requested):
- I'm a little confused with `examples/backup`. It seems the example supports creating a new postgres DB or passing an existing on in. If creating a new one, we seem to do a data lookup on the newly created instance in order to get a backup ID. However, when using an existing, we seem to ask for a postgres backup CRN. Which is it, ID or CRN? And do we even need to data lookup at all? Can we not just get the ID or CRN from the `postgresql_db` outputs? 
- Remove the `configuration` variable from the basic example
- Why is the complete example creating an ibm_iam_authorization_policy ? The main module itself supports doing this, so no need to do it in the example too
- With CBR support already added to this module, should we remove the deprecated `allowlist` feature (perhaps in a dubsequent PR)?
- Looks like we are missing some points in `examples/complete/README.md` - I see no mention of the VPE set up
- The title in `examples/complete/README.md` is not accurate. Suggest: Complete example with BYOK encryption, autoscaling, CBR rules, VPE creation and read-only replica provisioning
- Can you update CRA to use SCC v2 (and update the `examples/fscloud/README.md` to list any failing rules)

ocofaigh commented:
Oh also, when adding SCC v2 support, can you update the value for `TF_VAR_existing_at_instance_crn` to pull its value from the org variable `TF_VAR_existing_at_instance_crn=${{ vars.AT_INSTANCE_CRN }}`

rajatagarwal-ibm commented:
> * I'm a little confused with `examples/backup`. It seems the example supports creating a new postgres DB or passing an existing on in. If creating a new one, we seem to do a data lookup on the newly created instance in order to get a backup ID. However, when using an existing, we seem to ask for a postgres backup CRN. Which is it, ID or CRN? And do we even need to data lookup at all? Can we not just get the ID or CRN from the `postgresql_db` outputs?
> * Remove the `configuration` variable from the basic example
> * Why is the complete example creating an ibm_iam_authorization_policy ? The main module itself supports doing this, so no need to do it in the example too
> * With CBR support already added to this module, should we remove the deprecated `allowlist` feature (perhaps in a subsequent PR)?
> * Looks like we are missing some points in `examples/complete/README.md` - I see no mention of the VPE set up
> * The title in `examples/complete/README.md` is not accurate. Suggest: Complete example with BYOK encryption, autoscaling, CBR rules, VPE creation and read-only replica provisioning
> * Can you update CRA to use SCC v2 (and update the `examples/fscloud/README.md` to list any failing rules)

* We do need data lookup as the backup CRN has the following condition "backup_crn must be null OR starts with 'crn:' and contains ':backup:'" - which we are validating in the root module.
* Removed `configuration` from the basic example
* Yes, you are right `ibm_iam_authorization_policy` is not needed in the complete example as in the complete example we are using Key Protect Keys for the encryption (unlike FSCloud where we used HPCS)
* Yes, we can remove "allowlist" in the subsequent PR
* Updated README for the complete module, including title
* I will update to the SCCV2

rajatagarwal-ibm commented on tests/other_test.go:
```diff
@@ -8,6 +8,8 @@ import (
+const pitrTerraformDir = "examples/pitr"
+const basicExampleTerraformDir = "examples/basic"
 
 func TestRunRestoredDBExample(t *testing.T) {
```
So basically, we need to fetch the "backup crn" of the Postgres instance. The logic of the backup CRN is that we want to create a new Postgres instance based on the backup of the existing postgres instance.

Now we have 2 options:
1. Hardcode the backup CRN in the permanent resources file and point to that in the example. Although eventually that backup would be too old and might be automatically deleted (not sure about the timeline). 
2. Fetch backup similar to the way we do in the example. That might need some extra investigation. I can attempt in a separate PR if we want that.
3. Keep as is

ocofaigh commented on examples/complete/README.md:
```diff
@@ -1,10 +1,12 @@
 - Backend encryption using generated Key Protect key.
 - Create a Sample VPC.
 - Create Context Based Restriction(CBR) to only allow Postgresql to be accessible from the VPC.
+- Create a VPE.
```
Create a security group and a VPE for the postgres instance

ocofaigh reviewed (changes requested):
see final comments

terraform-ibm-modules-ops commented:
:tada: This PR is included in version 3.0.0 :tada:

The release is available on [GitHub release](https://github.com/terraform-ibm-modules/terraform-ibm-icd-postgresql/releases/tag/v3.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
